From 94a08adfd70a9a777ffb75bfe9aef2d17d681197 Mon Sep 17 00:00:00 2001
From: Tan Siewert <tan.siewert@9elements.com>
Date: Mon, 20 Jul 2026 22:59:37 +0200
Subject: [PATCH] build: allow compiling selected platforms only

platform-init links the init code for all platforms into the binary,
meaning it adds code for every board regardless of which one it targets.

On space-constrained systems this may not fit. A BMC with a 32MB SPI
(e.g. the HPE ProLiant Gen11 platforms) can barely fit all apps to run
the OpenBMC stack. Adding init code for platforms that will never run on
the target wastes valuable storage space.

Make the set of compiled-in platforms a build-time choice. A new
'platforms' meson array selects which platforms are compiled and
available. By default, all platforms will be selected to keep
backwards-compatibility. The init_functions array will be auto-generated
by meson via platform.hpp.in. If no platform is selected
(i.e. overwriting the default to be empty), or if platform-name is not
part of the selected platforms, an error will be thrown.

Building only nvidia-gb200 instead of all platforms can shrink the
compressed binary (stripped and splitted in squashfs.xz) by ~16 KiB:

    all platforms:    155648 bytes
    nvidia-gb200:     139264 bytes
    saving:            16384 bytes

Upstream-Status: Pending
Change-Id: I67f15741e1d5521ba55aa3b3c04d5af5b0b1641e
Signed-off-by: Tan Siewert <tan.siewert@9elements.com>
---
 meson.build     | 65 +++++++++++++++++++++++++++++++++++++++++++------
 meson.options   | 14 +++++++++++
 platform.cpp    | 17 +++----------
 platform.hpp.in | 16 ++++++++++++
 4 files changed, 90 insertions(+), 22 deletions(-)
 create mode 100644 platform.hpp.in

diff --git a/meson.build b/meson.build
index 2c8d240..8cb43b5 100644
--- a/meson.build
+++ b/meson.build
@@ -19,15 +19,55 @@ cli11_dep = dependency(
 )
 i2c_dep = meson.get_compiler('cpp').find_library('i2c')
 
-platform_srcs = files(
-    'platforms/intel/acrp.cpp',
-    'platforms/intel/jcrp.cpp',
-    'platforms/nvidia/gb200.cpp',
-    'platforms/nvidia/nvl32.cpp',
-    'platforms/nvidia/vr_nvl.cpp',
-)
 util_srcs = files('gpio.cpp', 'i2c.cpp', 'utilities.cpp')
 
+platform_name = get_option('platform-name')
+platforms = get_option('platforms')
+if platforms.length() == 0
+    error('at least one platform must be selected!')
+endif
+
+if platform_name not in platforms
+    error('selected platform name not selected in platforms')
+endif
+
+platforms_map = {
+    'intel-acrp': {
+        'src': files('platforms/intel/acrp.cpp'),
+        'fn': 'intel::init_acrp',
+    },
+    'intel-jcrp': {
+        'src': files('platforms/intel/jcrp.cpp'),
+        'fn': 'intel::init_jcrp',
+    },
+    'nvidia-gb200': {
+        'src': files('platforms/nvidia/gb200.cpp'),
+        'fn': 'nvidia::init_gb200_base',
+    },
+    'nvidia-gb200-with-p2020': {
+        'src': files('platforms/nvidia/gb200.cpp'),
+        'fn': 'nvidia::init_gb200_with_p2020',
+    },
+    'nvidia-nvl32': {
+        'src': files('platforms/nvidia/nvl32.cpp'),
+        'fn': 'nvidia::init_nvl32',
+    },
+    'nvidia-vr-nvl': {
+        'src': files('platforms/nvidia/vr_nvl.cpp'),
+        'fn': 'nvidia::init_vr_nvl',
+    },
+}
+
+platform_srcs = []
+platform_entries = []
+foreach platform : platforms
+    src = platforms_map[platform]
+    if src['src'] not in platform_srcs
+        platform_srcs += src['src']
+    endif
+    platform_entries += 'PLATFORM("@0@", @1@)'.format(platform, src['fn'])
+endforeach
+
 exe = executable(
     'platform',
     ['platform.cpp'] + platform_srcs + util_srcs,
@@ -44,6 +84,15 @@ configure_file(
     install_dir: systemd_system_unit_dir,
     install: true,
     configuration: configuration_data(
-        {'PLATFORM_NAME': get_option('platform-name')},
+        {'PLATFORM_NAME': platform_name},
     ),
 )
+
+conf = configuration_data()
+conf.set('PLATFORMS', ',\n    '.join(platform_entries))
+
+configure_file(
+    input: 'platform.hpp.in',
+    output: 'platform.hpp',
+    configuration: conf,
+)
diff --git a/meson.options b/meson.options
index 9004923..b30ff06 100644
--- a/meson.options
+++ b/meson.options
@@ -1,3 +1,17 @@
+option(
+    'platforms',
+    type: 'array',
+    choices: [
+        'intel-acrp',
+        'intel-jcrp',
+        'nvidia-gb200',
+        'nvidia-gb200-with-p2020',
+        'nvidia-nvl32',
+        'nvidia-vr-nvl',
+    ],
+    description: 'List of platforms to be included',
+)
+
 option(
     'platform-name',
     type: 'string',
diff --git a/platform.cpp b/platform.cpp
index 66663cb..4ca3729 100644
--- a/platform.cpp
+++ b/platform.cpp
@@ -1,10 +1,10 @@
 // SPDX-License-Identifier: Apache-2.0
 // SPDX-FileCopyrightText: Copyright OpenBMC Authors
 
+#include "platform.hpp"
+
 #include "gpio.hpp"
 #include "i2c.hpp"
-#include "intel/platforms.hpp"
-#include "nvidia/platforms.hpp"
 #include "utilities.hpp"
 
 #include <fcntl.h>
@@ -14,20 +14,9 @@
 #include <gpiod.hpp>
 
 #include <algorithm>
-#include <array>
-#include <iostream>
 #include <string_view>
 #include <utility>
 
-constexpr auto init_functions =
-    std::to_array<std::pair<std::string_view, int (*)()>>(
-        {{"intel-acrp", intel::init_acrp},
-         {"intel-jcrp", intel::init_jcrp},
-         {"nvidia-gb200", nvidia::init_gb200_base},
-         {"nvidia-gb200-with-p2020", nvidia::init_gb200_with_p2020},
-         {"nvidia-nvl32", nvidia::init_nvl32},
-         {"nvidia-vr-nvl", nvidia::init_vr_nvl}});
-
 int main(int argc, char** argv)
 {
     CLI::App app("Platform init CLI");
@@ -50,7 +39,7 @@ int main(int argc, char** argv)
         [&platform_name](const std::pair<std::string_view, int (*)()> val) {
             return val.first == platform_name;
         });
-    if (it == init_functions.end())
+    if (it == std::ranges::end(init_functions))
     {
         std::cerr << init_sub->help() << "\n";
         return EXIT_FAILURE;
diff --git a/platform.hpp.in b/platform.hpp.in
new file mode 100644
index 0000000..8759d29
--- /dev/null
+++ b/platform.hpp.in
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright OpenBMC Authors
+
+#pragma once
+
+#include "intel/platforms.hpp"
+#include "nvidia/platforms.hpp"
+
+#include <string_view>
+
+#define PLATFORM(name, func) {name, func}
+
+inline constexpr auto init_functions =
+    std::to_array<std::pair<std::string_view, int (*)()>>({
+    @PLATFORMS@
+});
